taf/testlib/common/loops.py

```diff
@@ -0,0 +1,303 @@
+"""
+@copyright Copyright (c) 2017, Intel Corporation.
+
+This program is free software; you can redistribute it and/or modify it
+under the terms and conditions of the GNU Lesser General Public License,
+version 2.1, as published by the Free Software Foundation.
+
+This program is distributed in the hope it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
+more details.
+
+@file  loops.py
+
+@summary  flexible and extensible conditional loops helpers.
+"""
+
+import time
+from contextlib import suppress
+from collections import OrderedDict
+
+
+class CondLoopException(Exception):
+    pass
+
+
+class CondLoopBase(object):
+    """
+    """
+    END_FAIL = False
+    END_SUCCESS = True
+    KEEP_LOOPING = None
+
+    def __init__(self, end_fail=False, end_success=True, keep_looping=None):
+        super(CondLoopBase, self).__init__()
+
+        self.END_FAIL = end_fail
```

Comment:
If we want instance attributes, then we should make instance attributes and not mask the class attributes.

`self.end_fail = end_fail`

By masking the class attributes, we could get strange behaviors as we move between `self` and `cls`.  `self.END_FAIL` should always be the same as `cls.END_FAIL`, but `self.end_fail` can be something else.

```
class A(object):
  VAL1 = 2

  def __init__(self, val1=VAL1):
    self.name1(self)
    self.name2(self)
    self.VAL1 = 999
    self.name1(self)
    self.name2(self)

  @classmethod
  def name1(cls, instance):
    print(' '.join('{}: {:10}'.format(k, str(v.VAL1)) for k, v in {'cls': cls, 'instance': instance}.items()))

  name2 = name1


a1 = A()
```
Outputs:
```
cls: 2          instance: 2
cls: 2          instance: 2
cls: 2          instance: 999
cls: 2          instance: 999
```